src/partition/repart.c

```diff
@@ -4884,17 +4884,24 @@ static int context_write_partition_table(Context *context) {
         if (r < 0)
                 return log_error_errno(r, "Failed to write partition table: %m");
 
-        capable = blockdev_partscan_enabled(fdisk_get_devfd(context->fdisk_context));
+        fd = fdisk_get_devfd(context->fdisk_context);
+
+        capable = blockdev_partscan_enabled(fd);
         if (capable == -ENOTBLK)
                 log_debug("Not telling kernel to reread partition table, since we are not operating on a block device.");
         else if (capable < 0)
                 return log_error_errno(capable, "Failed to check if block device supports partition scanning: %m");
         else if (capable > 0) {
                 log_info("Telling kernel to reread partition table.");
 
-                if (context->from_scratch)
+                if (context->from_scratch) {
+                        if (flock(fd, LOCK_EX) < 0)
+                                return log_error_errno(r, "Failed to lock block device: %m");
+
+                        CLEANUP_BSD_UNLOCK(fd);
+
                         r = fdisk_reread_partition_table(context->fdisk_context);
-                else
+                } else
                         r = fdisk_reread_changes(context->fdisk_context, original_table);
```

Comment:
I am not familiar with the difference, but why the locking is unnecessary for `fdisk_reread_changes()`?